internal/encryption/keyprovider/openbao/config.go

```diff
@@ -0,0 +1,97 @@
+package openbao
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	openbao "github.com/openbao/openbao/api"
+	"github.com/opentofu/opentofu/internal/encryption/keyprovider"
+)
+
+type Config struct {
+	Address string `hcl:"address,optional"`
+	Token   string `hcl:"token,optional"`
+
+	KeyName   string `hcl:"key_name"`
+	KeyLength int    `hcl:"key_length,optional"`
+}
+
+func (c Config) Build() (keyprovider.KeyProvider, keyprovider.KeyMeta, error) {
+	if c.KeyName == "" {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Message: "no key name found",
+		}
+	}
+
+	if c.KeyLength == 0 {
+		c.KeyLength = defaultKeyLength
+	}
+
+	if err := validateKeyLength(c.KeyLength); err != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Cause: err,
+		}
+	}
+
+	// DefaultConfig reads BAO_ADDR and some other optional env variables.
+	config := openbao.DefaultConfig()
+	if config.Error != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Cause: config.Error,
+		}
+	}
+
+	// Address from HCL supersedes BAO_ADDR.
+	if c.Address != "" {
+		config.Address = c.Address
+	}
+
+	client, err := newClient(config, c.Token)
+	if err != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Cause: err,
+		}
+	}
+
+	return &keyProvider{
+		svc:       service{client},
+		keyName:   c.KeyName,
+		keyLength: c.KeyLength,
+		ctx:       context.Background(),
+	}, new(keyMeta), nil
+}
+
+type clientConstructor func(config *openbao.Config, token string) (client, error)
+
+var errNoOpenBaoTokenFound = errors.New("no openbao token found")
+
+// newClient variable allows to inject different client implementations.
+// In order to keep client interface simple, token setting is in this function as well.
+// It's not possible to pass token in config.
+var newClient clientConstructor = func(config *openbao.Config, token string) (client, error) {
```

Comment:
This is a globally scoped injection, which makes parallelization of tests impossible and cleanup is needed after each test which is easy to mess up. I would prefer if we injected the constructor function directly, e.g. using a constructor function and provide a second constructor function that injects the defaults.

Reply:
Not sure that's what you expect, but I made it possible to reuse a default client constructor: [f460dfa](https://github.com/opentofu/opentofu/pull/1436/commits/f460dfa21c2affb99f90b4e8df97f1972e17d4fe). I think we still have a problem with tests parallelization in cases when we need multiple test suits run in parallel with different mocks / implementations. It is true for AWS and GCP providers as well. WDYT?